File: automator/src/main/java/com/stardust/automator/simple_action/SearchTargetAction.java
package com.stardust.automator.simple_action;

import android.graphics.BitmapFactory;

import com.stardust.automator.UiObject;

import java.util.List;

/**
 * Created by Stardust on 2017/1/27.
 */

public abstract class SearchTargetAction extends FilterAction {

    private int mAction;

    public SearchTargetAction(int action, Filter filter) {
        super(filter);
        mAction = action;
    }


    @Override
    public boolean perform(List<UiObject> nodes) {
        boolean performed = false;
        for (UiObject node : nodes) {
            node = searchTarget(node);
            if (node != null && performAction(node)) {
                performed = true;
            }
        }
        return performed;
    }

    protected boolean performAction(UiObject node) {
        return node.performAction(mAction);
    }

    public int getAction() {
        return mAction;
    }

    public UiObject searchTarget(UiObject node) {
        return node;
    }

    @Override
    public String toString() {
        return "SearchTargetAction{" +
                "mAction=" + mAction + ", " +
                super.toString() +
                "}";
    }
}

File: automator/src/main/java/com/stardust/automator/simple_action/SearchTargetAction.kt
package com.stardust.automator.simple_action

import android.graphics.BitmapFactory

import com.stardust.automator.UiObject

/**
 * Created by Stardust on 2017/1/27.
 */

abstract class SearchTargetAction(val action: Int, filter: FilterAction.Filter) : FilterAction(filter) {

    override fun perform(nodes: List<UiObject>): Boolean {
        var performed = false
        for (node in nodes) {
            val targetNode = searchTarget(node)
            if (targetNode != null && performAction(targetNode)) {
                performed = true
            }
        }
        return performed
    }

    protected open fun performAction(node: UiObject): Boolean {
        return node.performAction(action)
    }

    open fun searchTarget(node: UiObject?): UiObject? {
        return node
    }

    override fun toString(): String {
        return "SearchTargetAction{" +
                "mAction=" + action + ", " +
                super.toString() +
                "}"
    }
}
